README.md

```diff
@@ -1,6 +1,163 @@
+            return request
+
+        ontology_request = check_if_archivo_ontology_requested(request)
+        if ontology_request:
+            logger.debug('The request is for an ontology')
+            try:
+                ontology_url = str(request._url)
+                headers, _ = get_headers_and_expected_type(request)
+                response = requests.get(ontology_url, headers=headers)
+                if response.status_code == 502:
+                    logger.error('Received 502 Bad Gateway error')
+                    response = proxy_logic_http(request)
+                    logger.debug('Queue response')
+                    self.queue_response(response)
+                    return None
+                else:
+                    logger.debug('The request is correct')
+                    return request
+            except (SSLError, Timeout, ConnectionError, RequestException) as e:
+                logger.error(f'Network-related exception occurred {e}')
+                response = proxy_logic_http(request)
+                logger.debug('Queue response')
+                self.queue_response(response)
+                return None
+        return request
+
+
+    def handle_client_request(self, request: HttpParser):
+        logger.debug('HTTP call')
+        logger.debug(request._url)
+
+        ontology_request = check_if_archivo_ontology_requested(request)
+        if not ontology_request:
+            logger.info('No ontology is asked, forward original request')
+            return request    
+        
+        response = proxy_logic_http(request)
+        self.queue_response(response)
+
+        return None
+    
+
+    def handle_upstream_chunk(self, chunk: memoryview):
+        logger.info('HTTPS call')
+
+        try:
+            # Parse the HTTP response to handle different cases
+            parser = HttpParser(httpParserTypes.RESPONSE_PARSER)
+            parser.parse(memoryview(chunk))
+            code = int(parser.code.decode('utf-8'))
+            if code >= 100 and code < 200:
+                return chunk
+            elif code >= 201 and code <= 204:
+                return chunk
+            elif code == 451:
+                return chunk
+            else:
+                response = proxy_logic_https(parser)
+                logger.debug('Queue response')
+                self.queue_response(response)
+                return None
+        except UnicodeDecodeError:
+            logger.warning('Received non-text chunk, cannot decode')
+        except Exception as e:
+            logger.error(f'Exception occurred while handling upstream chunk: {e}')
+        
+        return chunk
+
+    def queue_response(self, response):
+        self.client.queue(
+            build_http_response(
+                response.status_code, 
+                reason=bytes(responses[response.status_code], 'utf-8'), 
+                headers={
+                    b'Content-Type': bytes(response.headers.get('Content-Type'), 'utf-8')
```

Comment:
please again carefully note that headers are typically not in utf-8
https://stackoverflow.com/a/4410331
if this is a conversion made proxy.py internally then please describe it and link it in a comment such that it uses python utf-8 strings internally and then converts them into asciii

Reply:
but it is best to actually test that by adding some weird unicode chars here and see what happens